torch/csrc/utils/tensor_new.cpp

```diff
@@ -216,8 +220,11 @@ Tensor internal_new_from_data(const Type & type, at::optional<Device> device_opt
                                                        *torch::getLayout(type.backend()),
                                                        type_inference_device_type);
       const auto& type_to_use = type_inference ? type_inference_type : type;
-      return copy_variables ? new_with_tensor_copy(type_to_use, var, device_index) :
+      auto new_tensor = copy_variables ? new_with_tensor_copy(type_to_use, var, device_index) :
                               new_with_type_conversion(type_to_use, var, device_index);
+      new_tensor.detach_(); // making copy constructed tensor a leaf node
```

Comment:
I'm pretty sure the `new_with_type_conversion` branch doesn't guarantee that it won't return the exact same variable (this is why we have the `copy_variables` flag!!), so this code is incorrect, and can modify the input to e.g. `torch.as_tensor`.

Reply:
Yes, this is incorrect.  I don't think passing `args_requires_grad` is the right mechanism here: it's not even checked in most of the branches, but there's no indication that this should be the case.  Just move the detach / set logic to where you need it (tensor_new and tensor_ctor).

Reply:
Thanks for pointing this out! Sorry I landed it without thinking it thoroughly. I will push a new PR to fix this